### PR TITLE
mdm 391

### PR DESCRIPTION
**Pull request by paxtonhare**
- adding ability to run custom actions

### CODE REVIEW

File: examples/dhf-flow/gradle-local.properties [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Put your overrides from gradle.properties here
+
+mlUsername=admin
```
Comment (dmcassel) [inline examples/dhf-flow/gradle-local.properties]: didn't we decide to get rid of this file?

File: examples/triggers/gradle-local.properties [hunk 2]
```diff
@@ -1,4 +1,4 @@
 # Put your overrides from gradle.properties here
 
-mlUsername=
-mlPassword=
+mlUsername=admin
```
Comment (dmcassel) [inline examples/triggers/gradle-local.properties]: checked in on purpose?

File: src/test/ml-modules/root/test/suites/merging-xml/custom-xqy-action.xqy [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+declare namespace sm = "http://marklogic.com/smart-mastering";
+
+(: Force update mode :)
+declare option xdmp:update "true";
+
+declare option xdmp:mapping "false";
+
+let $assertions :=
+  test:assert-false(xdmp:invoke-function(function() {
+    fn:doc-available("/sjs-action-output.json")
```
Comment (dmcassel) [inline src/test/ml-modules/root/test/suites/merging-xml/custom-xqy-action.xqy]: should this be /xqy-action-output.json?

File: examples/dhf-flow/gradle-local.properties (hunk 1 above)
Comment (paxtonhare) [reply examples/dhf-flow/gradle-local.properties]: doh!

File: src/main/ml-modules/root/com.marklogic.smart-mastering/impl/process.xqy [hunk 4]
```diff
@@ -95,6 +101,34 @@ declare function proc-impl:process-match-and-merge-with-options(
...
+    for $action in fn:distinct-values($actions)
+    let $action-xml := $matching-options//*:action[@name = $action]
+    let $action-func :=
+      fun-ext:function-lookup(
+        fn:string($action-xml/@function),
+        fn:string($action-xml/@namespace),
+        fn:string($action-xml/@at),
+        ()
       )
+    let $filtered-matches := $matching-results/result[@action = $action]
```
Comment (rjrudin) [inline src/main/ml-modules/root/com.marklogic.smart-mastering/impl/process.xqy]: @dmcassel @paxtonhare IIUC - this allows for a developer to provide a custom implementation for merging, correct? e.g. merge entity B into entity A, preserving the ID of entity A and not creating a new entity?

Comment (dmcassel) [reply src/main/ml-modules/root/com.marklogic.smart-mastering/impl/process.xqy]: yes, it allows for an arbitrary action. if someone implemented what you're suggesting, I wouldn't be confident about unmerging or auditing working as expected. The intent of this is to provide additional notification actions (send me an email) as opposed to changing the basic way things work, but that is a possibility

**Pull request merged**